src/persist/src/nemesis/generator.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use rand::prelude::SmallRng;
+use rand::{Rng, SeedableRng};
+
+use crate::nemesis::{
+    Input, ReadSnapshotReq, Req, ReqId, SealReq, SnapshotId, TakeSnapshotReq, WriteReq,
+};
+
+pub struct Generator {
+    seed: u64,
+    req_id: u64,
+    snap_id: u64,
+}
+
+// TODO: Make Generator configurable, so individual tests can be written to
+// stress interesting combinations of random traffic.
+impl Generator {
+    pub fn from_seed(seed: u64) -> Self {
+        Generator {
+            seed,
+            req_id: 0,
+            snap_id: 0,
+        }
+    }
+
+    pub fn gen(&mut self) -> Input {
+        let req_id = ReqId(self.req_id);
+        self.req_id += 1;
+        let mut rng = SmallRng::seed_from_u64(self.seed + req_id.0);
+        let req = loop {
+            break match rng.gen_range(0..=3) {
+                0 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let key = rng.gen_range('a'..'e').to_string();
+                    let ts = rng.gen_range(0..100);
+                    let diff = if rng.gen_bool(0.9) { 1 } else { -1 };
+                    Req::Write(WriteReq {
+                        stream,
+                        update: ((key, ()), ts, diff),
+                    })
+                }
+                1 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let ts = rng.gen_range(0..100);
+                    Req::Seal(SealReq { stream, ts })
+                }
+                2 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let snap = SnapshotId(self.snap_id);
+                    self.snap_id += 1;
+                    Req::TakeSnapshot(TakeSnapshotReq { stream, snap })
```

Comment:
Are you allowed to have any other operations over snapshots besides Taking and then restoring one (e.g. delete it, compact it, whatever)? The test does not recycle snapshot ids, so I assume Mz in production will also not be recycling snapshot ids as it is running?

Reply:
The snapshot IDs are actually just an artifact of the structure of the nemesis test (and one of the things I meant to call out when I mentioned "I still owe some comments in the code" in my request for your review). A snapshot user just gets an in-memory object which it uses directly, but I wanted to stress race conditions between snapshots and varies write/compaction/etc operators as well as eventually the edge cases around the still TODO disk-backed LRU cache for the persisted data. The snapshot IDs are what I came up with to do this while keeping Generator deterministic.